docs: Refactor ngmodule-api.md to make it more useful (#38258)
angular/angular: aio/content/guide/ngmodule-api.md

> # Using the NgModule metadata

Reviewer: I suppose you haven't done the nav yet, but I'm sure "NgModule metadata" will be better in the TOC that "NgModule API" 😃

Author: Included the navigation.json change in this PR in the most recent commit.

An [NgModule](guide/glossary#ngmodule "Definition of NgModule") is a class marked by the `@NgModule` decorator with a metadata object that describes how that particular part of the app fits together with the other parts.

## Prerequisites

This topic requires a basic understanding of the following:

* [Typescript](guide/glossary#typescript "Definition of Typescript") and HTML5 programming
* [Templates](guide/glossary#template "Definition of a template") in HTML with CSS styles
* [Components](guide/glossary#component "Definition of a component")

> * [Angular CLI](cli "Angular CLI")

Author: I'm using this same "template" set of bullets in the Creating an NgModule tutorial. In this case the CLI does make an appearance, so knowledge of it is needed because I don't explain it.

## Prerequisites

This topic requires a basic understanding of the following:

* [Typescript](guide/glossary#typescript "Definition of Typescript") and HTML5 programming
* [Templates](guide/glossary#template "Definition of a template") in HTML with CSS styles
* [Components](guide/glossary#component "Definition of a component")
* [Angular CLI](cli "Angular CLI")
* [NgModules](guide/ngmodules "NgModules")

## How the metadata guides compilation

The `@NgModule` metadata provides important information for the Angular compilation process that converts the app code you write into highly performant JavaScript code.
The metadata provides the following arrays:

* [`declarations`](#declarations): Which [components](guide/glossary#component "Definition of component"), [directives](guide/glossary#directive "Definition of directive"), and [pipes](guide/glossary#pipe "Definition of pipe)") belong to the NgModule.
  These classes are called [declarables](guide/glossary#declarable "Definition of a declarable").

> * [`entryComponents`](#entry-components): A list of [entry components](guide/entry-components "Entry components") that can be dynamically loaded into the view.

Author: I'm sorry but I learned internally at Google to use colons, and the style guide also reinforces that: [Colons instead of dashes in lists](https://developers.google.com/style/dashes#colons-instead-of-dashes-in-lists).

> * [`entryComponents`](#entry-components): A list of [entry components](guide/entry-components "Entry components") that can be dynamically loaded into the view.

Reviewer 2: Yes, I just saw that myself and have asked Dave how he'd like us to handle it. As the docs are currently, em dashes are all over. We could switch them as we go (as in this PR), switch them in a sweep, or possibly stay with the em dashes, or something else I haven't thought of. I'll let you know what I hear back.

For more information, see [Entry components](guide/entry-components "Entry components").

<div class="callout is-helpful">
<header>Why a component appears in both arrays</header>

`AppComponent` is listed in both the `declarations` and `bootstrap` arrays for the `AppModule` root NgModule.
You might see the same component listed in `exports` and `entryComponents`.
While that seems redundant, these metadata arrays have different functions.
Membership in one doesn't imply membership in another.
For example, you might declare `AppComponent` in your NgModule but not use it for bootstrapping, or `AppComponent` might be declared in a different NgModule.

</div>

{@a providers-array}

## The `providers` array

The `providers` array is a list of [dependency-injection](guide/glossary#dependency-injection-di "Definition of dependency injection") providers that describes the services that the app needs, and how to create an [injector](guide/glossary#injector "Definition of injector") at runtime for each service.
Angular registers the providers you include in this array with the NgModule's injector.

> If you include a provider in the _root_ NgModule's `providers` array, it becomes the root injector.

Reviewer 2: ```suggestion
If you include a provider in the _root_ NgModule's `providers` array, its service becomes visible to the root `ModuleInjector`.
```
The provider, service, and injectors are each different things (and there are different injectors). @jelbourn might like to fact check my edits here. I could be missing some nuance in the wording.
  
This is what the current doc says:
>Angular registers these providers with the NgModule's injector. If it is the NgModule used for bootstrapping then it is the root injector.

"Registering with the NgModule's injector" and" if it's used for bootstrapping" are key points. The passive voice in that second sentence obscures the idea that "it" means the `ModuleInjector`, not that "it"=the provider.

> Services for providers declared in the root become available for injection into any component, directive, pipe, or service which is a child of this injector.

Reviewer 2: ```suggestion
Services you configure for the root become available for injection into any component, directive, pipe, or service which is a child of this injector.
```

That same sentence that's currently in the doc that may be helpful for context.
>Angular registers these providers with the NgModule's injector.

I think there can be some important nuance between "available", "visible", "registered", and "configured". I hope @jelbourn can check me here.🙏🏼

> You can scope service usage when [creating other NgModules](guide/feature-modules "Feature modules").

Reviewer 2: ```suggestion
You can scope service visibility when [creating other NgModules](guide/feature-modules "Feature modules").
```

A [lazy-loaded NgModule](guide/lazy-loading-ngmodules "Lazy-loading feature modules") has its own injector, which is typically a child of the root injector.
Lazy-loaded services are scoped to the lazy-loaded NgModule's injector.
If a lazy-loaded NgModule also provides the `UserService`, any component created within that NgModule's context (such as by router navigation) gets the local instance of the service, not the instance in the root injector.
Components in external NgModules continue to receive the instance provided by their injectors.

To learn more about service providers, see [Providing dependencies in modules](guide/providers "Providing dependencies in modules").

{@a bootstrap-array}

## The `bootstrap` array

The `bootstrap` array is list of components that Angular automatically bootstraps to start your app.

> The Angular CLI command `ng new` generates the root NgModule for a new app project, as described in [Launching your app with a root module](guide/bootstrapping "Launching your app with a root module").

Reviewer 2: `ng new` generates the whole project, which includes the `AppModule`. We want to make sure that `ng new` doesn't sound like the command for generating just the root NgModule. How about something like:
```suggestion
The Angular CLI command `ng new` generates an Angular project, which includes the root NgModule by default. 
For more information, see  [Launching your app with a root module](guide/bootstrapping "Launching your app with a root module").
```

The app starts by launching the [entry component](guide/entry-components "Entry components") in the root NgModule's `bootstrap` array, which is `AppComponent`.

> A common practice is to use this `AppComponent` as the only entry component in this array.

Reviewer 2: Rather than a common practice, it is specifically the default in a CLI-generated app. So, though it's the default, you can do other things (as you describe below). With the CLI, most people won't configure this.

{@a imports}

## The `imports` array

The NgModule's `imports` array describes other NgModules needed for this NgModule.
The `imports` array accepts only `@NgModule` references.
The other NgModules must export the needed components, directives, or pipes that this NgModule needs.

The other NgModules' exported components, directives, and pipes are treated as if they were declared in this NgModule.
You can use them in this NgModule's templates.
For example, your component can use the `NgIf` and `NgFor` directives if your NgModule has imported `CommonModule` (perhaps indirectly by importing `BrowserModule`).

{@a exports}

## The `exports` array

The `exports` array is a list of [declarables](guide/glossary#declarable "Definition of a declarable") that another NgModule can use if it imports this NgModule.

For example, if an NgModule includes `UserComponent` in its `exports` array, components and templates in other NgModules can use `UserComponent` as long as the other NgModules import this NgModule.
If you don't include `UserComponent` in the `exports` array, only components and templates within this NgModule can use it.

> Export only declarable classes (components, directives, or pipes).

Reviewer 2: ```suggestion
Export only declarable classes; that is, only use the exports array for components, directives, or pipes.
```
Parentheses reduction. Here's an alternative route that uses a semicolon, but that requires an independent clause. Another option would be:

>Only use the exports array for components, directives, or pipes&mdash;collectively referred to as declarables.

But that uses the passive. Anyway, ideas to help avoid ().

You can export any declarable declared in this NgModule or in an imported NgModule.

## Re-exporting an entire NgModule

Importing an NgModule does _not_ automatically re-export the imported NgModule's imports.

> For example, NgModule B can't use `ngIf` just because it imported NgModule A, which imported `CommonModule` in order to includer `ngIf`.

Reviewer 2: ```suggestion
For example, NgModule B can't use `ngIf` just because it imports NgModule A, which imports `CommonModule` in order to include `ngIf`.
```

Sticking to present tense as much as possible.

Module B must also import `CommonModule`.

However, you can add another NgModule to your NgModule's `exports` array so that all of the other NgModule's public components, directives, and pipes are _re-exported_ by your NgModule.
If NgModule A re-exports `CommonModule` and NgModule B imports NgModule A, NgModule B components can use `ngIf` even though NgModule B didn't import `CommonModule` directly.

An NgModule can export a combination of its own declarations, selected imported classes, and imported NgModules.
You can selectively aggregate classes from other NgModules and re-export them in a consolidated, convenience NgModule.
For example, Angular's own `BrowserModule` exports a couple of NgModules as follows:

```typescript
  exports: [CommonModule, ApplicationModule]
```

Don't bother re-exporting pure service NgModules, because they don't export declarables that another NgModule could use.
For example, there's no point in re-exporting `HttpClientModule` because it doesn't export anything.
For more information about creating a service NgModule, see [Providing dependencies in modules](guide/providers "Providing dependencies in modules").

> ## Next steps

Reviewer 2: ```suggestion
## What's next
```

})
```

{@a declarations}

## The declarations array

The `declarations` array identifies the [declarables](guide/glossary#declarable "Definition of a declarable") ([components](guide/glossary#component "Definition of component"), [directives](guide/glossary#directive "Definition of directive"), and [pipes](guide/glossary#pipe "Definition of pipe)")) that belong to the NgModule.

All of the declarables for an NgModule must be in the NgModule's `declarations` array.

> You can use the [Angular CLI](cli "Angular CLI") to [generate](cli/generate) a new component, directive, or pipe in your NgModule, so that the CLI automatically adds the declarable to the NgModule's `declarations` array.

Reviewer: When you use the `ng generate` [CLI command](cli/generate) to create a new component, directive, or pipe in your NgModule, the CLI automatically adds...

Author: Fixed in next commit.

Each component, directive, and pipe that you include in an NgModule's `declarations` array belongs _only_ to that NgModule.

*Do not* include the following in the `declarations` array:

* A class that's already declared in another NgModule, another [Javascript module](guide/ngmodule-vs-jsmodule "JavaScript modules vs. NgModules"), or a third-party module or NgModule.
  The compiler emits an error if you try to declare the same class in more than one NgModule.
  For example, don't declare `FORMS_DIRECTIVES` from `@angular/forms` because the `FormsModule` already declares it.
  Be careful not to re-declare a class that is imported directly or indirectly from another NgModule.

> * NgModule or service classes, or non-Angular classes and objects, such as strings, numbers, functions, entity models, configurations, business logic, or helper classes.

Reviewer: I'd like to see some more about why no service classes.  What if this NgModule includes a service class definition?
Perhaps if the NgModule defines a service class, it includes the provider of that service in the the `providers` array, instead of declaring the service class itself -- is that right?
Or perhaps the service class is an export?

Author: After reading [Providing dependencies](https://angular.io/guide/providers#providing-dependencies-in-modules), I think the best example and description for providing a service is there, so I will add a link to it. I will also reference it back where we talk about declarables. 

Thanks!

Author: I'm editing the last bullet point to the following:

> * NgModule or service classes, or non-Angular classes and objects, such as strings, numbers, functions, entity models, configurations, business logic, or helper classes.
> Service providers belong in the [`providers`](#providers-array).
> For more information about creating a service NgModule, see [Providing dependencies in modules](guide/providers "Providing dependencies in modules").

{@a entry-components}

## The entryComponents array

The `entryComponents` array describes the [entry components](guide/entry-components "Specifying an entry component") that you can dynamically load into the view.

> If you use the CLI [new](cli/new "ng new") command to create an initial Angular app, the CLI creates one entry component, the root component called `AppComponent`, in the root NgModule.

Reviewer: When you use the `ng new` [CLI command](cli/new) to create...

Author: Fixed in next commit.

This component serves as a point of entry into the app.
You [bootstrap](guide/glossary#bootstrap "Definition of bootstrap") it to launch the app, as described in the section about the [bootstrap array](#bootstrap-array).

[Routing components](guide/glossary#routing-component "Definition of routing component") are also entry components because they are loaded dynamically by the router.
The router creates them and drops them into the DOM near a `<router-outlet>`.
You don't have to add the routing components to an NgModule's `entryComponents` array because Angular adds them implicitly.

> The only components you need to add to the `entryComponents` array are any components you are bootstrapping using one of the imperative techniques, such as [`ViewComponentRef.createComponent()`](api/core/ViewContainerRef#createComponent), as undiscoverable.

Reviewer: What does "undiscoverable" mean? by whom? in what circumstances? why?

While using the router is common in most apps, other dynamic component loading is not common.

> If you need to dynamically load components, you must add these components to the `entryComponents` array.

Reviewer: This is not at all clear - maybe just include a more general ref to the other page? Something like "In some cases, you might need to load another entry component dynamically. For more information..."

Author: Fixed in next commit.

Lazy-loaded services are scoped to the lazy-loaded NgModule's injector.
If a lazy-loaded NgModule also provides the `UserService`, any component created within that NgModule's context (such as by router navigation) gets the local instance of the service, not the instance in the root injector.
Components in external NgModules continue to receive the instance provided by their injectors.

To learn more about service providers, see [Providing dependencies](guide/providers "Providing dependencies for an NgModule").

{@a bootstrap-array}

## The bootstrap array

The `bootstrap` array is list of components that Angular automatically bootstraps to start your app.

The Angular CLI command `ng new` generates the root NgModule for a new app project, as described in [Launching an app with a root NgModule](guide/bootstrapping "Launching an app with a root NgModule").
The app starts by launching the [entry component](guide/entry-components "Entry components") in the root NgModule's `bootstrap` array, which is `AppComponent`.

Your app typically has only this one component in this list.

> However, Angular can launch with multiple bootstrap components, each with its own location in the host web page.

Reviewer: Is there some reason to do this?

Author: Rewrote as follows:

> A common practice is to use this `AppComponent` as the only entry component in this array.
> Angular creates the component and inserts it into the browser DOM for the host web page.
> 
> However, Angular can launch an app with multiple bootstrap components, each with its own location in the host web page.
> Angular creates each component that appears in the `bootstrap` array and inserts it into the browser DOM.
> Each bootstrapped component is the base of its own tree of components.
> Inserting a bootstrapped component usually triggers a cascade of component creations that fill out that tree.
> Any component you add to `bootstrap` is automatically added to `entryComponents`.
>

Angular creates each component that appears in the `bootstrap` array and inserts it into the browser DOM.
Each bootstrapped component is the base of its own tree of components.
Inserting a bootstrapped component usually triggers a cascade of component creations that fill out that tree.
Any component you add to `bootstrap` is automatically added to `entryComponents`.

{@a imports}

## The imports array

The NgModule's `imports` array describes other NgModules needed for this NgModule.
The `imports` array accepts only `@NgModule` references.
The other NgModules must export the needed components, directives, or pipes that this NgModule needs.

The other NgModules' exported components, directives, and pipes are treated as if they were declared in this NgModule.

> You can therefore can use a component, directive, or pipe in this NgModule's templates.

Reviewer: typo - but perhaps just "You can use them in this NgModule's templates."

Author: Fixed in the next commit.

> For example, your component can use the `NgIf` and `NgFor` directives if your NgModule has imported `CommonModule` (perhaps indirectly by importing `BrowserModule`).

Reviewer: not perhaps, this always happens.
"For example, your components can use the `NgIf` and `NgFor` directives because your NgModule has imported `CommonModule` (indirectly, by importing `BrowserModule`).

Author: It happens because AppModule imports BrowserModule. CommonModule imports and then re-exports the BrowserModule stuff. But devs creating new modules should import CommonModule to get these directives, not BrowserModule, so leaving it out the parenthetical as it's not necessary. Fixed in next commit.

{@a exports}

## The exports array

> The `exports` array is a list of declarations that another NgModule can use if the other NgModule imports this NgModule.

Reviewer: "...if it imports..."?

Author: Fixed in next commit.

Export only [declarable](guide/glossary#declarable "Definition of a declarable") classes (components, directives, or pipes) that you want other NgModules to be able to use in their templates.
You can export any declarable whether it's declared in this NgModule or in an imported NgModule.
The declarables you export are your NgModule's _public_ classes.
If you don't export a declarable class, it stays _private_, visible only to other components declared in this NgModule.

For example, if an NgModule includes `UserComponent` in its `exports` array, components and templates in other NgModules can use `UserComponent` as long as the other NgModules import this NgModule.

> If you don't include `UserComponent` in the `exports` array, only components and templates within this NgModule can use it.

Reviewer: I don't really see that this example adds anything.

Author: I changed this section to use the example and cut out repetition. Here's the revised section:

> ## The exports array
> 
> The `exports` array is a list of [declarables](guide/glossary#declarable "Definition of a declarable") that another NgModule can use if it imports this NgModule.
> 
> For example, if an NgModule includes `UserComponent` in its `exports` array, components and templates in other NgModules can use `UserComponent` as long as the other NgModules import this NgModule.
> If you don't include `UserComponent` in the `exports` array, only components and templates within this NgModule can use it.
> 
> Export only declarable classes (components, directives, or pipes).
> You can export any declarable declared in this NgModule or in an imported NgModule.
>

> ## Re-exporting an entire NgModule

Reviewer: h3, I think

Author: Decided to keep this h2 because (1) it's a topic by itself (even though you use the exports array), and (2) there is no other possible h3 to go along with it, and I hate to have only one subhead under a heading.

Importing an NgModule does _not_ automatically re-export the imported NgModule's imports.
For example, NgModule B can't use `ngIf` just because it imported NgModule A, which imported `CommonModule` in order to includer `ngIf`.
Module B must also import `CommonModule`.

However, you can add another NgModule to your NgModule's `exports` array so that all of the other NgModule's public components, directives, and pipes are _re-exported_ by your NgModule.

> If NgModule A re-exports `CommonModule` and NgModule B imports NgModule A, NgModule B components can use `ngIf` even though NgModule B didn't import `CommonModule`.

Reviewer: ...didn't import `CommonModule` directly.

Author: Fixed in next commit.

An NgModule can export a combination of its own declarations, selected imported classes, and imported NgModules.
You can selectively aggregate classes from other NgModules and re-export them in a consolidated, convenience NgModule.
For example, Angular's own `BrowserModule` exports a couple of NgModules as follows:

```typescript
  exports: [CommonModule, ApplicationModule]
```

> Don't bother re-exporting pure service NgModules, because they don't export declarables that another NgModule could use.

Reviewer: Oh! this is interesting - maybe this could be mentioned back where we talked about declarables.

Is there any other topic here that considers the case of an NgModule that defines a service? It doesn't look like it, but maybe I missed it. It's not the same as "Providing dependencies for an NgModule" -- kind of the opposite, in fact.

Author: After reading [Providing a service](https://angular.io/guide/providers#providing-a-service), I think "service NgModule" is wrong -- NgModules contain references to service providers, but services themselves are individual classes (individual files) in the workspace. `ng generate service User` generates `UserService` in `src/app/user.service.ts`. That means, to me, that the service is _not_ inside the NgModule. You cannot therefore have a "service NgModule". This is cloudy because you can have a _service module_ in the strict JavaScript sense -- `src/app/user.service.ts` is a service module (a file) in JavaScript.

I will use some of [Providing a service](https://angular.io/guide/providers#providing-a-service) back where we talk about declarables. Thanks!

Author: This seems to be true: [`HttpClientModule`](https://angular.io/api/common/http/HttpClientModule) is an example of an NgModule that provides a service (that's all it does). The actual service is [HttpClient](https://angular.io/api/common/http/HttpClient).
